WalletWasabi/Crypto/ZeroKnowledge/LinearRelation/Statement.cs

```diff
@@ -0,0 +1,51 @@
+using NBitcoin.Secp256k1;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using WalletWasabi.Helpers;
+
+namespace WalletWasabi.Crypto.ZeroKnowledge.LinearRelation
+{
+	public class Statement
+	{
+		public Statement(params Equation[] equations)
+			: this(equations as IEnumerable<Equation>)
+		{
+		}
+
+		public Statement(IEnumerable<Equation> equations)
+		{
+			// The equation matrix should not be jagged
+			Guard.NotNullOrEmpty(nameof(equations), equations);
+			var n = equations.First().Generators.Count();
+			Guard.True(nameof(equations), equations.All(e => e.Generators.Count() == n));
+
+			foreach (var generator in equations.SelectMany(equation => equation.Generators))
+			{
+				Guard.NotNull(nameof(generator), generator);
+			}
+
+			Equations = equations;
+		}
+
+		public IEnumerable<Equation> Equations { get; }
+
+		public bool CheckVerificationEquation(GroupElementVector publicNonces, Scalar challenge, IEnumerable<ScalarVector> allResponses)
+		{
+			// The responses matrix should match the generators in the equations and
+			// there should be once nonce per equation.
+			Guard.True(nameof(publicNonces), Equations.Count() == publicNonces.Count());
+			Equations.CheckDimesions(allResponses);
+
+			return Equations.Zip(publicNonces, allResponses, (equation, r, s) => equation.Verify(r, challenge, s)).All(x => x);
+		}
+
+		public GroupElementVector SimulatePublicNonces(Scalar challenge, IEnumerable<ScalarVector> allGivenResponses)
```

Comment:
Why do you actually use `IEnumerable<ScalarVector>` here? 

I would guess that you actually need a data structure that has `Count` property and that you can index using `[]` so that you don't need those zips.

`ICollection<ScalarVector>` extends `IEnumerable<T>` and it has `Count`. `IList<T>` extends `IEnumerable<T>` and it has `[]` defined.

Reply:
For readability/understandability i personally prefer `Zip` over a loop with indexing, but defer to others. Although maybe i'm misunderstanding the suggestion on how to use `[]`?

see also this comment thread about an alternative approach: https://github.com/zkSNACKs/WalletWasabi/pull/4280#discussion_r482639719

Reply:
I'll ask on Slack about this.

Reply:
As we discussed offline, I think arrays are the best in this case. However we will continue with this as it is and make sure we pass arrays from the top level API objects. We can test/improve the different data structures incrementally.